### DISCUSSION TITLE
Invariant Violation: Failed to publish: the publish script exited.

### DISCUSSION THREAD

**Discussion by leo-petrucci** [ossjs/release#47] [Q&A] [answered]
I've used Release before without issues, but I'm trying to publish a new package and it's just failing without much help.

Repo can be found [here](https://github.com/creativiii/meow-cms).

```
❯ yarn run release
INFO: preparing release for "creativiii/meow-cms" from branch "main"...
INFO: found latest release: 0.0.5 (80f52735548b9dc4ebdb66caabca6b18bd9026a1)
INFO: found 1 new commit:
  - 5e53aeb60bb14b5a486225191fd121577dd5e4b4 fix: 🔧 Fix release config (#37)
INFO: successfully parsed 1 commit(s)!
INFO: release type "patch": 0.0.5 -> 0.0.6
INFO: bumped version in package.json to:
INFO: preparing to run the publishing script with:
{"RELEASE_VERSION":"0.0.6"}
INFO: executing publishing script...
cli.sh publish [options]

Options:
      --version  Show version number                                   [boolean]
      --help     Show help                                             [boolean]
  -d, --dry-run  Print command steps without executing them
                                                      [boolean] [default: false]

Invariant Violation: Failed to publish: the publish script exited.
    at Publish.runReleaseScript (/Users/leonardo/meow-cms/node_modules/@ossjs/release/bin/build/commands/publish.js:154:36)
    at processTicksAndRejections (internal/process/task_queues.js:95:5)
    at async Publish.run (/Users/leonardo/meow-cms/node_modules/@ossjs/release/bin/build/commands/publish.js:100:9)
```

The package publishes just fine with `yarn npm publish` so I'm not too sure why it's not working through Release.

**Comment by kettanaito** [comment]
Hey, @creativiii. Could you try the latest Release version? I've implemented std streaming from the script process to the Release process so you should get stderr/stdout now, giving you more details on what's going wrong. This is 99% an error thrown by the release script (missing registry auth, improper `yarn` CLI usage, etc). I've been bitten by this quite a number of times, thus the streaming feature. 

Give the latest version a try and let me know.

**Reply by leo-petrucci** [reply]
Last version was `0.4.0` released 5 months ago according to npm. Is there a newer one?

**Reply by kettanaito** [reply]
Thanks for pointing this out. I've spotted that the automated release job has been failing because I forgot to build the package before releasing itself. Should be fixed in #48.

**Reply by kettanaito** [reply]
I've released the latest state of the library as [0.5.0](https://github.com/ossjs/release/releases/tag/v0.5.0). Please update to it in your setup and see the release output for any error indications. Those should now be streamed correctly.

**Reply by leo-petrucci** [reply]
Sorry for the late reply. I installed `0.5.0`, tried releasing, didn't work. So I removed my lockfile and cleared my deps to start fresh, tried releasing again but still didn't work.

I'm still getting the same error without specifics :/
```
Invariant Violation: Failed to publish: the publish script exited.
    at Publish.runReleaseScript (/Users/leonardo/meow-cms/node_modules/@ossjs/release/bin/build/commands/publish.js:170:36)
    at processTicksAndRejections (internal/process/task_queues.js:95:5)
    at async Publish.run (/Users/leonardo/meow-cms/node_modules/@ossjs/release/bin/build/commands/publish.js:106:9)
```

[Updated branch](https://github.com/creativiii/meow-cms/tree/ci/update-release-dep)

**Reply by kettanaito** [reply]
Can you try the release locally in dry mode? Use the `-d` (`--dry-run`) option for the Release CLI to have a dry run. You'd have to provide your GITHUB/NPM tokens anyway because Release reads from your repo even in the dry run, just doesn't write anything.

**Reply by leo-petrucci** [reply]
Yeah it's always worked fine in dry run mode:
```
yarn run release                                              
INFO: preparing release for "creativiii/meow-cms" from branch "main"...
INFO: found latest release: 0.0.5 (80f52735548b9dc4ebdb66caabca6b18bd9026a1)
INFO: found 14 new commits:
  - 052ea95043bfb307f289b0cf67ac9942f76e5c54 Merge branch 'main' of github.com:creativiii/meow-cms
  - 0ab0d38f5e8595abbc24006d5b62626c9a2b1cce feat: ✨ Added links to editor (#39)
  - 0f81c8847c674dce43c641bf9d11656742619a36 Updated "imported-ironeko-article" from MeowCMS
  - 65e2a8dd5377f41699b1611b2644847374d331be Updated "Donec odio" from MeowCMS
  - 4dc7658b693400b7e73b82475e499878189ab495 Updated "Donec odio" from MeowCMS
  - 633a86d7b554031b4df38f9abe043aeb0f6629c7 Working links
  - bddaefd05cf196fffae65d24360a9b26896ee5df Updated "Donec odio" from MeowCMS
  - 4312b573f30e49d3a669e6d0e97585fe7c5082fc Updated "Donec odio" from MeowCMS
  - b53d18a591d5f9b76ef1d5267f7aac2220001717 Updated "Donec odio" from MeowCMS
  - b84e4669a4f731f0b304d07ae7a12b4ca93b91d7 Updated "Donec odio" from MeowCMS
  - 8e4c45b9f0d3d6e17f6aba39b667fb6741d7405a Updated "Donec odio" from MeowCMS
  - 0189393ce617917361f203a26f47676e8cf489c0 Updated "Donec odio" from MeowCMS
  - 667d2d406da64677f0a396127c932cc8c481f739 Updated "Donec odio" from MeowCMS
  - 5e53aeb60bb14b5a486225191fd121577dd5e4b4 fix: 🔧 Fix release config (#37)
INFO: successfully parsed 14 commit(s)!
INFO: release type "minor": 0.0.5 -> 0.1.0
WARN: skip version bump in package.json in dry-run mode (next: 0.1.0)
INFO: preparing to run the publishing script with:
{"RELEASE_VERSION":"0.1.0"}
WARN: skip executing publishing script in dry-run mode
WARN: skip creating a release commit in dry-run mode: "chore(release): v0.1.0"
WARN: skip creating a release tag in dry-run mode: v0.1.0
WARN: skip pushing release to Git in dry-run mode
INFO: generated release notes:

## v0.1.0 (2023-02-15)

### Features

- ✨ Added links to editor (#39) (0ab0d38f5e8595abbc24006d5b62626c9a2b1cce) @creativiii

### Bug Fixes

- 🔧 Fix release config (#37) (5e53aeb60bb14b5a486225191fd121577dd5e4b4) @creativiii

WARN: skip creating a GitHub release in dry-run mode
WARN: skip commenting on 2 GitHub issues:
  - 39
  - 37
WARN: release "v0.1.0" completed in dry-run mode!
```

That's why this is so weird 😂

**Reply by kettanaito** [reply]
Figured... Since the release script is not run in dry mode, the dry mode will succeed. 

It's strange that you're not getting any stderr output on CI. If the release script writes to stderr/stdout and/or exits, you should be able to see the reason (given the script reports one). 

The next thing I'd suggest is to run that release script standlone (maybe it supports a dry option as well! depends on what you're using). So what I mean is taking this:

```js
// ossjs.release.config.js
module.exports = { script: 'COMMAND' }
```

And running that `COMMAND` separately (preferably on CI so we get closer to the environment). If that script throws it will also throw standalone.

**Comment by leo-petrucci** [comment, marked as answer]
It just dawned on me that I have no reason to use `yarn` to publish the package and that I can just use `npm publish` to send the package to NPM. 

So I ended up doing that and it worked really easily. I think yarn 2 might've changes some stuff in regards to the publish command so it might be worthy of investigation. Two things I noticed:

- It complained about using `--new-version`
- Once I removed it running `yarn npm publish` started failing with a `404` error, even though I've used the exact command before to publish it locally

Either way, Release now runs fine, so all good! Thank you for help and sorry for wasting your time!

**Reply by kettanaito** [reply]
It's still odd that it never forwarded those errors. Sometimes CLIs manage stderr poorly, writing to stdout instead, maybe that was the case. 

Anyhow, there's been quite some battle-testing when it comes to using Release with different package managers, so I highly recommend you looking into what we describe in README:

- [Usage with Yarn](https://github.com/ossjs/release#usage-with-yarn)
- [Usage with PNPM](https://github.com/ossjs/release#usage-with-pnpm)

(Usage with NPM is the most straightforward one since NPM will publish any current state of the package).

**Reply by leo-petrucci** [reply]
Yeah that's what I was following, but I've got a snaking suspicion that the command is for `yarn v1`, and it doesn't work with v2.

Specifically:
- `--new-version` I think doesn't exist anymore
- the command is now `yarn npm publish`

Even after removing those I then wasn't able to get Release to run properly, and even just running `yarn npm publish` in my terminal (which had previously worked) started throwing 404 errors. So I decided to simply give up on it.

**Reply by kettanaito** [reply]
True! I wrote those docs for the tools I'm using and I've never used Yarn v2 yet. Could you open a PR to README to clearly indicate that section concerns Yarn V1 only? To prevent any confusion for others. 

Also, if you've cracked how to make it work with V2, we can add that to the README as well.
